compiler/src/iree/compiler/Codegen/Common/GPU/GPUVectorDistribution.cpp

```diff
@@ -0,0 +1,272 @@
+#include "iree/compiler/Codegen/Common/GPU/VectorLayoutProvider.h"
+#include "iree/compiler/Codegen/Common/VectorLayoutAnalysis.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/Vector/IR/VectorOps.h"
+#include "mlir/IR/Verifier.h"
+
+#define DEBUG_TYPE "iree-codegen-gpu-vector-distribution"
+
+using namespace mlir::iree_compiler::IREE::VectorExt;
+
+namespace mlir::iree_compiler {
+
+TypedValue<VectorType> getDistributed(RewriterBase &rewriter,
+                                      TypedValue<VectorType> value,
+                                      LayoutProvider *provider) {
+  // If this is a result of a "to_simd" op, use the source value of it.
+  if (auto toSIMD = value.getDefiningOp<IREE::VectorExt::ToSIMDOp>()) {
+    value = cast<TypedValue<VectorType>>(toSIMD.getInput());
+    return value;
+  }
+  // Create a "to_simt" op to convert the value to the distributed layout.
+  SmallVector<int64_t> distributedShape = provider->getDistributedShape(value);
+  VectorType distributedType =
+      VectorType::get(distributedShape, value.getType().getElementType());
+  auto toSIMT = rewriter.create<IREE::VectorExt::ToSIMTOp>(
+      value.getLoc(), distributedType, value);
+  return toSIMT.getResult();
+}
+
+void replaceOpWithDistributedValues(RewriterBase &rewriter, Operation *op,
+                                    LayoutProvider *provider,
+                                    ValueRange values) {
+  // Replace all OpResults with the given values.
+  SmallVector<Value> replacements;
+  for (OpResult opResult : op->getOpResults()) {
+    Value replacement = values[opResult.getResultNumber()];
+    // If this value is a vector type, it must be converted back to simd.
+    if (isa<VectorType>(replacement.getType())) {
+      auto oldResult = cast<TypedValue<VectorType>>(opResult);
+      // Create a toSIMD op to convert the value back to the simd.
+      rewriter.setInsertionPointAfterValue(oldResult);
+      auto toSIMD = rewriter.create<IREE::VectorExt::ToSIMDOp>(
+          oldResult.getLoc(), oldResult.getType(), replacement);
+      // Clone the layout to the new value.
+      provider->getAnalysis().cloneLayoutInformationToNewValue(
+          oldResult, toSIMD.getResult());
+      // Add to replacements.
+      replacement = toSIMD.getResult();
+    }
+    replacements.push_back(replacement);
+  }
+
+  rewriter.replaceOp(op, replacements);
+}
+
+class DistributionRewriter : public IRRewriter, public RewriterBase::Listener {
+public:
+  DistributionRewriter(MLIRContext *ctx, DenseSet<Operation *> &erasedOps,
+                       SmallVector<Operation *> &worklist,
+                       VectorLayoutAnalysis &analysis)
+      : IRRewriter(ctx), erasedOps(erasedOps), worklist(worklist),
```

Comment:
we can let the caller decide what small size to use
```suggestion
                       SmallVectorImpl<Operation *> &worklist,
                       VectorLayoutAnalysis &analysis)
      : IRRewriter(ctx), erasedOps(erasedOps), worklist(worklist),
```